### PR TITLE
CYB-79 Integration tests for  CDAP flume sink

### PR DESCRIPTION
**Pull request by ytolochkevych**

### CODE REVIEW

File: it/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+CDAP Ingest Integration Tests
+==================
```
Comment (awholegunch) [inline it/README.md]: Underlines not used in Markdown. Use
# Cask Data Application Platform (CDAP) Ingest Integration Tests

File: it/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+This project contains set of integration tests for various CDAP ingest project components.
+Tests are executed against running CDAP instance. Connection information is configured in properties files.
+This files are stored in resources directory per each sub-module. You can launch same integration tests with different
```
Comment (awholegunch) [inline it/README.md]: These files are stored in the resource directory in each sub-module. You can launch the same integration tests with different configuration file using maven profiles: it-local, it-local-auth, it-remote and it-remote-auth.

File: it/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+==================
+
+This project contains set of integration tests for various CDAP ingest project components.
+Tests are executed against running CDAP instance. Connection information is configured in properties files.
+This files are stored in resources directory per each sub-module. You can launch same integration tests with different 
+configuration file using maven profiles: it-local, it-local-auth, it-remote and it-remote-auth
+
+## Usage
+
+ To launch integration tests against local CDAP instance execute:
```
Comment (awholegunch) [inline it/README.md]: To launch integration tests against Standalone CDAP instance, execute:

File: it/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ 
+ ```
+ mvn clean install -P it-local
+ ```
+
+ To launch integration tests against local CDAP instance with authentication enabled, execute:
```
Comment (awholegunch) [inline it/README.md]: To launch integration tests against Standalone CDAP instance with authentication enabled, execute:

File: it/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ 
+ ```
+ mvn clean install -P it-local-auth
+ ```
+ 
+ To launch integration tests against remote CDAP instance execute:
```
Comment (awholegunch) [inline it/README.md]: To launch integration tests against Distributed CDAP instance, execute:

File: it/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ 
+ ```
+ mvn clean install -P it-remote
+ ```
+
+ To launch integration tests against remote CDAP instance with authentication enabled, execute:
```
Comment (awholegunch) [inline it/README.md]: To launch integration tests against Distributed CDAP instance with authentication enabled, execute:

File: it/cdap-flume-java-integration-tests/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
+CDAP Flume Sink Integration Tests
```
Comment (awholegunch) [inline it/cdap-flume-java-integration-tests/README.md]: # Cask Data Application Platform (CDAP) Flume Sink Integration Tests

File: it/cdap-flume-java-integration-tests/README.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+==================
+
+This project contains integration tests for CDAP Flume Sink
```
Comment (awholegunch) [inline it/cdap-flume-java-integration-tests/README.md]: This project contains integration tests for CDAP Flume Sink.

File: it/cdap-flume-java-integration-tests/README.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+## Usage
+
+ Please review general CDAP Ingest IT documentation for usage instructions. This document contains information about
```
Comment (awholegunch) [inline it/cdap-flume-java-integration-tests/README.md]: Please review the general [CDAP ingest documentation](docs.cask.co/cdap/current) for usage instructions. This document contains information about project-specific configuration files.

File: it/cdap-flume-java-integration-tests/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ project specific configuration files.
+
+## Configuration
+
+ To configure integration tests against local CDAP instance edit:
```
Comment (awholegunch) [inline it/cdap-flume-java-integration-tests/README.md]: To configure integration tests against Standalone CDAP instance, edit:

File: it/cdap-flume-java-integration-tests/README.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ 
+ ```
+ src/test/resources/local.conf 
+ ```
+
+ To configure integration tests against local CDAP instance with authentication enabled, edit:
```
Comment (awholegunch) [inline it/cdap-flume-java-integration-tests/README.md]: To configure integration tests against Standalone CDAP instance with authentication enabled, edit:

File: it/cdap-flume-java-integration-tests/README.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+ 
+ ```
+ src/test/resources/local_auth.conf 
+ ```
+ 
+ To configure integration tests against remote CDAP instance edit:
```
Comment (awholegunch) [inline it/cdap-flume-java-integration-tests/README.md]: To configure integration tests against Distributed CDAP instance, edit:

File: it/cdap-flume-java-integration-tests/README.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+ 
+ ```
+ src/test/resources/remote.conf 
+ ```
+
+ To configure integration tests against remote CDAP instance with authentication enabled, edit:
```
Comment (awholegunch) [inline it/cdap-flume-java-integration-tests/README.md]: To configure integration tests against Distributed CDAP instance with authentication enabled, edit:

File: it/stream-client-java-integration-tests/README.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
+CDAP Stream Client Integration Tests
```
Comment (awholegunch) [inline it/stream-client-java-integration-tests/README.md]: # Cask Data Application Platform (CDAP) Stream Client Integration Tests

File: it/stream-client-java-integration-tests/README.md [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+==================
+
+This project contains integration tests for CDAP Stream Client Java API.
```
Comment (awholegunch) [inline it/stream-client-java-integration-tests/README.md]: This project contains integration tests for the CDAP Stream Client Java API.

File: it/stream-client-java-integration-tests/README.md [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+## Usage
+
+ Please review general CDAP Ingest IT documentation for usage instructions. This document contains information about
```
Comment (awholegunch) [inline it/stream-client-java-integration-tests/README.md]: Please review the general [CDAP ingest documentation](docs.cask.co/cdap/current) for usage instructions. This document contains information about project-specific configuration files.

File: it/stream-client-java-integration-tests/README.md [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+ project specific configuration files.
+
+## Configuration
+
+ To configure integration tests against local CDAP instance edit:
```
Comment (awholegunch) [inline it/stream-client-java-integration-tests/README.md]: To configure integration tests against Standalone CDAP instance, edit:

File: it/stream-client-java-integration-tests/README.md [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+ 
+ ```
+ src/main/resources/local.conf 
+ ```
+
+ To configure integration tests against local CDAP instance with authentication enabled, edit:
```
Comment (awholegunch) [inline it/stream-client-java-integration-tests/README.md]: To configure integration tests against Standalone CDAP instance with authentication enabled, edit:

File: it/stream-client-java-integration-tests/README.md [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+ 
+ ```
+ src/main/resources/local_auth.conf 
+ ```
+ 
+ To configure integration tests against remote CDAP instance edit:
```
Comment (awholegunch) [inline it/stream-client-java-integration-tests/README.md]: To configure integration tests against Distributed CDAP instance, edit:

File: it/stream-client-java-integration-tests/README.md [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+ 
+ ```
+ src/main/resources/remote.conf 
+ ```
+
+ To configure integration tests against remote CDAP instance with authentication enabled, edit:
```
Comment (awholegunch) [inline it/stream-client-java-integration-tests/README.md]: To configure integration tests against Distributed CDAP instance with authentication enabled, edit:

File: it/stream-client-java-integration-tests/README.md [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+ 
+ ```
+ src/main/resources/remote_auth.conf 
+ ```
+ 
+ Please refer CDAP Stream Client Java API for additional documentation.
```
Comment (awholegunch) [inline it/stream-client-java-integration-tests/README.md]: Please refer to the [CDAP Stream Client Java API](docs.cask.co/cdap/current) for additional documentation.

**Comment by awholegunch** [comment]
I've added comments on the documentation.

**Comment by ashvayka** [comment]
All doc issues fixed

**Comment by anew** [comment]
The structure is not clear to me. Do we now have a separate maven project "it" that has a module for each test? And we run maven with different profiles to run the tests? How does this integrate with our existing test automation (audi, primer)? I think this needs a larger discussion that will span other integration tests, too.

File: it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+
+    agent.configure(propertyMap);
+    agent.start();
+    for (int i = 0; i < eventNumber; i++) {
+      Event event = new SimpleEvent();
+      agent.put(event);
+    }
+      agent.stop();
+  }
+  @Test(expected = EventDeliveryException.class)
```
Comment (anew) [inline it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java]: missing newline before this line

File: it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+      propertyMap.put(name, properties.getProperty(name));
+    }
+
+    agent.configure(propertyMap);
+    agent.start();
+    for (int i = 0; i < eventNumber; i++) {
+      Event event = new SimpleEvent();
+      agent.put(event);
+    }
+      agent.stop();
```
Comment (anew) [inline it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java]: indentation is off.

File: it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+
+  @Before
+  public void setUP() throws Exception {
+    Field field = EmbeddedAgentConfiguration.class.getDeclaredField("ALLOWED_SINKS");
+    field.setAccessible(true);
+
+    Field modifiersField = Field.class.getDeclaredField("modifiers");
+    modifiersField.setAccessible(true);
+    modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL);
+    String []sinkList = {"co.cask.cdap.flume.StreamSink"};
```
Comment (anew) [inline it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java]: whitespace after [] and not before?

File: it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+    agent.stop();
+  }
+
+
+  @Before
+  public void setUP() throws Exception {
+    Field field = EmbeddedAgentConfiguration.class.getDeclaredField("ALLOWED_SINKS");
+    field.setAccessible(true);
+
+    Field modifiersField = Field.class.getDeclaredField("modifiers");
```
Comment (anew) [inline it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java]: this is a bit ugly. Is this the only way you can configure the allowed sinks?

File: it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java (hunk 23 above)
Comment (anew) [inline it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java]: this test needs to verify that the events were actually written, and that they are in correct order.

File: it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java (hunk 22 above)
Comment (ytolochkevych) [inline it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java]: ok

File: it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java (hunk 23 above)
Comment (ytolochkevych) [inline it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java]: Exception is throwing  in agent.put method if cdap sink fail to process event

File: it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java (hunk 25 above)
Comment (ytolochkevych) [inline it/cdap-flume-java-integration-tests/src/test/java/co/cask/cdap/flume/CdapFlumeIT.java]: EmbeddedAgentConfiguration class has private field with string array of the allowed sinks. For test is simplest way to add cdap sink to allowed

**Comment by ytolochkevych** [comment]
I have made some changes. Please review

**Comment by ashvayka** [comment]
@anew We don't have any audi/primer tests in this repo. The request was to provide integration tests that we are able to launch with different configuration files. We have provided IT and 4 maven profiles to launch tests: local, local with auth, remote and remote with auth. Let's have a separate discussion on reorganizing project structure. I believe we need more accurate requirements to do this.

**Comment by ashvayka** [comment]
This PR depends on https://github.com/caskco/cdap-ingest/pull/26. We need to merge auth client integration and then cover this integration in current PR.

**Comment by anew** [comment]
I like the new structure better. We will figure out later how we can integrate this with Primer.

I do still think that a test must verify its success. After sending events using flume, the test should also verify that these events are actually in the stream and in the correct order. Without that the test is not complete.

**Comment by ytolochkevych** [comment]
@anew.  I did not find any way  how to get  information from the  stream about event processing, number of event etc. If it is possible, please,  send any info how do it

**Comment by anew** [comment]
@ytolochkevych the least you can do is use the stream API to read the events back from the stream. And you can also retrieve metrics about the stream through the REST API. I am not sure whether metrics is sufficient, though. Especially because StreamClient is asynchronous, you want to verify that the events were written - and can be read - in the correct order.

**Comment by anew** [comment]
http://docs.cdap.io/cdap/2.5.0/en/api.html#reading-events-from-a-stream
http://docs.cdap.io/cdap/2.5.0/en/api.html#metrics-http-api

**Comment by anew** [comment]
@ytolochkevych any updates on this?

**Comment by ytolochkevych** [comment]
@anew  yes,  there are updates. I have created new pull request: https://github.com/caskdata/cdap-ingest/pull/56, because repo was re-forked

**Comment by prinam** [comment]
@anew can you pls verify that all your comments were addressed in #56 and close this PR. Thank you!

**Pull request closed without merging**